include/fiction/io/read_fgl_layout.hpp

```diff
@@ -0,0 +1,813 @@
+                            incoming_signals.push_back(incoming_signal_i);
+                        }
+                        kitty::dynamic_truth_table tt_t(static_cast<uint32_t>(num_incoming_signals));
+                        kitty::create_from_hex_string(tt_t, gate.type);
+                        lyt.create_node({incoming_signals}, tt_t, location);
+                    }
+                }
+                else
+                {
+                    throw fgl_parsing_error(fmt::format("Error parsing FGL file: unknown gate type: {}", gate.type));
+                }
+            }
+        }
+
+        return lyt;
+    }
+
+  private:
+    /**
+     * The layout which will be altered based on the parsed information.
+     */
+    Lyt lyt;
+    /**
+     * The input stream from which the gate-level layout is read.
+     */
+    std::istream& is;
+    /**
+     * @struct gate_storage
+     *
+     * Represents a gate in a fcn layout, storing its unique ID, type, name, location, and incoming connections.
+     */
+    struct gate_storage
+    {
+        /**
+         * Unique identifier for the gate.
+         */
+        int id{};
+        /**
+         * Type of the gate, can be an alias (AND, OR, PI, ..) or the implemented function in a binary or hexadecimal
+         * form.
+         */
+        std::string type;
+        /**
+         * Name of the gate (for inputs and outputs).
+         */
+        std::string name;
+        /**
+         * Location of the gate represented its x-, y- and z-coordinate.
+         */
+        tile<Lyt> loc{};
+        /**
+         * List of incoming connections to the gate.
+         */
+        std::vector<tile<Lyt>> incoming{};
+
+        /**
+         * Static member function to compare gate_storage objects by their IDs.
+         *
+         * @param gate1 First gate to be compared.
+         * @param gate2 Second gate to be compared.
+         * @return True if gate1's ID is less than gate2's ID, false otherwise.
+         */
+        static bool compare_by_id(const gate_storage& gate1, const gate_storage& gate2) noexcept
+        {
+            return gate1.id < gate2.id;
+        }
+    };
+};
+
+}  // namespace detail
+
+/**
+ * Reads a gate-level layout from an FGL file provided as an input stream.
+ *
+ * May throw a `fgl_parsing_exception` if the FGL file is malformed.
+ *
+ * @tparam Lyt The layout type to be created from an input.
+ * @param is The input stream to read from.
+ * @param name The name to give to the generated layout.
+ */
+template <typename Lyt>
+[[nodiscard]] Lyt read_fgl_layout(std::istream& is, const std::string_view& name = "")
+{
+    static_assert(is_coordinate_layout_v<Lyt>, "Lyt is not a coordinate layout");
+    static_assert(is_tile_based_layout_v<Lyt>, "Lyt is not a tile-based layout");
+    static_assert(is_clocked_layout_v<Lyt>, "Lyt is not a clocked layout");
+    static_assert(is_gate_level_layout_v<Lyt>, "Lyt is not a gate-level layout");
```

Comment:
I might be wrong, but I think that the `is_gate_level_layout` trait includes the clocked, tile-based, and coordinate layout traits.